enhancements/microshift/update-and-rollback.md

```diff
@@ -0,0 +1,180 @@
+via the `ostree-finalize-staged.service` systemd unit.
+
+Furthermore, in order to apply this policy, the `rpm-ostreed-automatic.service` systemd
+service must be enabled.
+
+#### MicroShift's Upgrade Systemd Service
+
+A new systemd service file will be packaged into the MicroShift's main RPM. This systemd
+service will monitor the `/run/ostree/staged-deployment` file that stores the staged update deployment info by using the `"ConditionPathExists"` option. This condition will be the
+trigger to run a custom script that saves the MicroShift state into a data backup file.
+
+#### MicroShift Backup
+
+Leveraging the ability to run any custom script before the new OSTree deployment becomes
+active, allow us to perform the following tasks in order:
+* Stop MicroShift systemd service (which does not cause interruption for applications)
+* Save a tarball file of `/var/lib/microshift` to `/var/lib/microshift-backup`
+* Save snapshots of any persistent information store in volumes
+* Reboot to get the new OSTree deployment enabled
+
+#### MicroShift Rollback
+
+`Greenboot` will perform different checks to verify if the Edge Device is in a healthy state.
+If the result is declared as successful, MicroShift will run normally until a new OSTree
+version is exposed to the device. If the restult is declared as failed, `Greenboot` will
+run custom scripts that will perform the following tasks in order:
+* Stop MicroShift systemd service
+* Restore all data from `/var/lib/microshift-backup` directory
+* Execute `rpm-ostree rollback -r` in order to deploy the previous N-1 state
+
+### Risks and Mitigations
+
+MicroShift does not run the Cluster Version Operator to handle upgrades, and it relies
+on OSTree capabilities for transactional updates and rollbacks. There could be ocasions
+when specific tasks are required to perform a successful upgrade of certain components
+such as deprecated API removal, ETCD major versions bump, etc. In order to mitigate this,
+every MicroShift version will have a specific custom upgrade script to jump to the next
+version.
+
+MicroShift is meant to run on Edge Devices with small footprint. Storing backups could
+lead the device to run out of disk space. To mitigate this, we will only support only
+N-1 states, which translates into saving only the backup of a previous state and not
+several of them.
+
+As mentioned in the MicroShift Greenboot enhancement, if network connectivity is slow or
+unstable, it can lead to a failed state which will end up in a rollback. In this case,
+increasing timers or reinstalling the upgrade when network improves will be necessary.
+
+MicroShift depends on OpenVswitch for its networking infrastructure. The ovn-master should maintain the database schema in sync with the specific OVN version.
+
+### Drawbacks
+
+The way MicroShift is upgraded differs a lot from classic OpenShift. The lack of Cluster
+Operators allow the small footprint of MicroShift while it prevents a more standard way of
```

Comment:
Bigger picture though, I can definitely imagine us using greenboot/rollbacks even in clustered/datacenter OpenShift too.  I think it just hasn't had someone to champion/drive it.